### PR TITLE
List backups action

### PR DESCRIPTION
**Pull request by javierdelapuente**
Applicable spec: ISD095

### Overview

This PR creates the action `list-backups` action in Synapse. 

The backups created with the action `create-backup` will be created in a specific prefix in the S3 repository, and they will be read and listed by this action, with some extra information. Its output will be like this:

```
juju run synapse/leader list-backups
Running operation 51 with 1 task
  - task 52 on unit-synapse-0

Waiting for task 52...
backups:
  backup-20240206100503602202:
    last-modified: 2024-02-06 10:05:05.945000+00:00
    size: "40699"
  backup-20240206152530632850:
    last-modified: 2024-02-06 15:25:33.826000+00:00
    size: "40698"
  backup-20240206152707269036:
    last-modified: 2024-02-06 15:27:09.728000+00:00
    size: "40699"
  backup-20240207074717616279:
    last-modified: 2024-02-07 07:47:18.803000+00:00
    size: "40699"
formatted: |-
  backup-id                     | last-modified                |            size
  ------------------------------------------------------------------------------
  backup-20240206100503602202   | 2024-02-06 10:05:05.945000+00:00 |           40699
  backup-20240206152530632850   | 2024-02-06 15:25:33.826000+00:00 |           40698
  backup-20240206152707269036   | 2024-02-06 15:27:09.728000+00:00 |           40699
  backup-20240207074717616279   | 2024-02-07 07:47:18.803000+00:00 |           40699
```

`list-backups` will iterate over the objects in S3 at the given path. As elements will not be retrieved and checked, if a 
`fake` backup is put in S3 in that path, it will be listed.
 
### Rationale

This new action is part of the disaster and recovery purposes related to backups.

### Module Changes

Added `list-backups` action. Added new observer in `backup_observer` and logic to query S3 in `backup` module.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Manging Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: src/backup.py [hunk 1]
```diff
@@ -161,6 +176,46 @@ def can_use_bucket(self) -> bool:
...
+    def list_backups(self) -> List[S3Backup]:
+        """List the backups stored in S3 in the current s3 configuration.
+
+        Returns:
+            list of backups.
+        """
+        backups = []
+        for item in self._iterate_objects():
+            s3_object_key = pathlib.Path(item["Key"])
+            backup_id = (s3_object_key).relative_to(self._prefix)
```
Comment (arturo-seijas) [inline src/backup.py:188]: Why the parathesis?

File: src/backup.py [hunk 2]
```diff
@@ -161,6 +176,46 @@ def can_use_bucket(self) -> bool:
...
+            backup = S3Backup(
+                backup_id=str(backup_id),
+                last_modified=item["LastModified"],
+                size=item["Size"],
+            )
+            backups.append(backup)
+        return backups
+
+    def _iterate_objects(self) -> Generator[dict, None, None]:
```
Comment (arturo-seijas) [inline src/backup.py:197]: I find the method name a big generic

File: src/backup.py (hunk 1 above)
Comment (javierdelapuente) [reply src/backup.py]: no reason at all. Removed!

File: src/backup.py (hunk 2 above)
Comment (javierdelapuente) [reply src/backup.py]: Renamed to `list_s3_objects` that I believe gives it a hint that we are querying S3

File: src/backup_observer.py [hunk 3]
```diff
@@ -94,3 +96,56 @@ def _on_create_backup_action(self, event: ActionEvent) -> None:
...
+        except ValueError:
+            logger.exception("Wrong S3 configuration in list backups action")
+            event.fail("Wrong S3 configuration on list backups action. Check S3 integration.")
+            return
+
+        try:
+            s3_client = backup.S3Client(s3_parameters)
+            backups = s3_client.list_backups()
+        except backup.S3Error:
+            logger.exception("Error listing backups.")
```
Comment (arturo-seijas) [inline src/backup_observer.py:136]: It would be nice to log the exception message for debugging purposes

File: src/backup_observer.py [hunk 4]
```diff
@@ -94,3 +96,56 @@ def _on_create_backup_action(self, event: ActionEvent) -> None:
...
+    def _on_list_backups_action(self, event: ActionEvent) -> None:
+        """List backups in S3 configured storage.
+
+        Args:
+            event: Event triggering the list backups action.
+        """
+        try:
+            s3_parameters = backup.S3Parameters(**self._s3_client.get_s3_connection_info())
+        except ValueError:
+            logger.exception("Wrong S3 configuration in list backups action")
```
Comment (arturo-seijas) [inline src/backup_observer.py:128]: It would be nice to log the exception message for debugging purposes

File: src/backup.py (hunk 2 above)
Comment (arturo-seijas) [reply src/backup.py]: Thanks

File: src/backup_observer.py (hunk 3 above)
Comment (javierdelapuente) [reply src/backup_observer.py:136]: By default, the exception (including the trace) is added with the `logger.exception`, see https://docs.python.org/3/library/logging.html#logging.Logger.exception

File: src/backup_observer.py (hunk 4 above)
Comment (javierdelapuente) [reply src/backup_observer.py:128]: By default, the exception (including the trace) is added with the logger.exception, see https://docs.python.org/3/library/logging.html#logging.Logger.exception

**Review by arturo-seijas — DISMISSED** [review]

File: src/backup_observer.py [hunk 5]
```diff
@@ -94,3 +96,56 @@ def _on_create_backup_action(self, event: ActionEvent) -> None:
             return
 
         event.set_results({"result": "correct", "backup-id": backup_id})
+
+    def _generate_backup_list_formatted(self, backup_list: typing.List[backup.S3Backup]) -> str:
```
Comment (cbartz) [inline src/backup_observer.py:100]:
```suggestion
    def _generate_backup_list_formatted(self, backup_list: list[backup.S3Backup]) -> str:
```

File: src/backup.py [hunk 6]
```diff
@@ -161,6 +176,46 @@ def can_use_bucket(self) -> bool:
             return False
         return True
 
+    def list_backups(self) -> List[S3Backup]:
```
Comment (cbartz) [inline src/backup.py:179]:
```suggestion
    def list_backups(self) -> list[S3Backup]:
```

**Review by cbartz — DISMISSED** [review]
LGTM, nitpicks only

**Review by cbartz — DISMISSED** [review]

**Review by cbartz — APPROVED** [review]

**Review by arturo-seijas — APPROVED** [review]

**Pull request merged**